internal/tar_interpreter_new.go

```diff
@@ -85,3 +93,34 @@ func createLocalFile(targetPath, name string) (*os.File, error) {
 	}
 	return file, nil
 }
+
+func (tarInterpreter *FileTarInterpreter) applyFileUnwrapResult(result *FileUnwrapResult, fileName string) {
+	switch result.FileUnwrapResultType {
+	case Skipped:
+		return
+	case Completed:
+		tarInterpreter.addToCompletedFiles(fileName)
+	case CreatedFromIncrement:
+		tarInterpreter.addToCreatedPageFiles(fileName, result.blockCount)
+	case WroteIncrementBlocks:
+		tarInterpreter.addToWrittenIncrementFiles(fileName, result.blockCount)
+	}
+}
+
+func (tarInterpreter *FileTarInterpreter) addToCompletedFiles(fileName string) {
+	tarInterpreter.mutex.Lock()
```

Comment:
1. You use the same lock for independent fields. It's slow.
2. In such a case you should use channels. (I remember I said that mutex is faster, and it's true in lots of cases. Here channel and mutex have similar performance, but channels are more convenient way)

Reply:
Fixed in 3880c5da3d1ab954ef87a0bc380b27e9a3992fce